### PR TITLE
ghc 9 support

### LINKED ISSUE #69: ghc 9 - Couldn't match MonadCatch with MonadThrow
MonadCatch hasn't been touched since 2019 so this probably has something to do with ghc 9 changes.
```hs
[ 3 of 31] Compiling Polysemy.ConstraintAbsorber.MonadCatch
                            
/home/theo/prj/polysemy-zoo/src/Polysemy/ConstraintAbsorber/MonadCatch.hs:83:20: error:
    * Couldn't match type: C.MonadCatch (Sem r)
                     with: C.MonadThrow (Sem r)
      Expected: (C.MonadThrow (Sem r) => Sem r a) -> Sem r a
        Actual: (C.MonadCatch (Sem r) => Sem r a) -> Sem r a
    * In the expression: absorbMonadCatch
      In an equation for `absorbMonadThrow':
          absorbMonadThrow = absorbMonadCatch
    * Relevant bindings include
        absorbMonadThrow :: (C.MonadThrow (Sem r) => Sem r a) -> Sem r a
          (bound at src/Polysemy/ConstraintAbsorber/MonadCatch.hs:83:1)
   |                        
83 | absorbMonadThrow = absorbMonadCatch
   |                    ^^^^^^^^^^^^^^^^
```
The `stack.yaml` I used:
```yml
compiler: ghc-9.0.1
resolver: nightly-2020-05-07

allow-newer: true

packages:
- .

extra-deps:
- atomic-primops-0.8.4
- compact-0.2.0.0
- random-1.2.0
- splitmix-0.1.0.3
- syb-0.7.2.1
- th-abstraction-0.4.2.0
- QuickCheck-2.14.2
# GHC 9 patch (not yet on hackage)
- github: polysemy-research/polysemy
  commit: 09c65bdae9c9ce50a55c7ee2ab9681916f09f4c3
  subdirs: [.]
```

### PR DESCRIPTION
**Pull request by funketh**
I inlined `absorbMonadCatch` and changed the type application which fixed #69.
Should probably still look into if/why this is necessary.

### CODE REVIEW

**Comment by funketh** [comment]
now getting this error which is most likely related to the ghc 9 changes of `Data.Coerce`:
```hs
[18 of 31] Compiling Polysemy.ConstraintAbsorber.MonadCont

/home/theo/prj/polysemy-zoo/src/Polysemy/ConstraintAbsorber/MonadCont.hs:31:14: error:
    * Couldn't match type: forall b1. a1 -> Sem r b1
                     with: a1 -> Sem r b
      Expected: ((a1 -> Sem r b) -> Sem r a1) -> Sem r a1
        Actual: ((forall b. a1 -> Sem r b) -> Sem r a1) -> Sem r a1
    * In the first argument of `ContDict', namely `(callCC @ref)'
      In the third argument of `absorbWithSem', namely
        `(ContDict (callCC @ref))'
      In the expression:
        absorbWithSem
          @C.MonadCont @Action (ContDict (callCC @ref)) (Sub Dict)
    * Relevant bindings include
        absorbCont :: (C.MonadCont (Sem r) => Sem r a) -> Sem r a
          (bound at src/Polysemy/ConstraintAbsorber/MonadCont.hs:30:1)
   |
31 |   (ContDict (callCC @ref))
   |              ^^^^^^^^^^^
```

**Comment by funketh** [comment]
the new error seems to be related to #66

**Comment by funketh** [comment]
could have something to do with the [minor changes](https://downloads.haskell.org/ghc/9.0.1/docs/html/users_guide/9.0.1-notes.html#language) to type variable quantification in ghc 9 (I don't see how though...)

**Are we sure that the current version compiled before ghc 9?**
(I'm getting a ghc panic with the repo's `stack.yaml`, will have to try a different snapshot)

**Comment by KingoftheHomeless** [comment]
The new error is definitely a victim of the loss of function contravariance. Violent eta-expansion should fix it.
```haskell
(ContDict (\main -> callCC @ref $ \exit -> main exit))
```

**Comment by funketh** [comment]
now builds with ghc 9 for me

File: src/Polysemy/ConstraintAbsorber/MonadCont.hs [hunk 1]
```diff
@@ -28,7 +28,7 @@ absorbCont
        -- 'Sem'. This might be something with type @'C.MonadCont' m => m a@.
     -> Sem r a
 absorbCont = absorbWithSem @C.MonadCont @Action
-  (ContDict (callCC @ref))
+  (ContDict (\main -> callCC @ref $ \exit -> main exit))
```
Comment (KingoftheHomeless) [inline src/Polysemy/ConstraintAbsorber/MonadCont.hs:31]: Out of curiosity, does simply `(ContDict (\main -> callCC @ref main))` work? My guess is no.

**Review by KingoftheHomeless — APPROVED** [review]

File: src/Polysemy/ConstraintAbsorber/MonadCont.hs (hunk 1 above)
Comment (funketh) [reply src/Polysemy/ConstraintAbsorber/MonadCont.hs:31]: nope, can't match `a1 -> Sem r b` with `forall b1. a1 -> Sem r b1`.

Comment (KingoftheHomeless) [reply src/Polysemy/ConstraintAbsorber/MonadCont.hs:31]: Suspected as much.

**Pull request merged**